### PR TITLE
onFlush and postFlush Entity Listeners

### PR DESCRIPTION
**Pull request by alexgurrola**
Simply check if an onFlush or postFlush entity listener is subscribed
for the current class in the entity manager during dispatch.

This is a rebased followup to #5822, per @lcobucci's review.

### CODE REVIEW

File: lib/Doctrine/ORM/UnitOfWork.php [hunk 1]
```diff
@@ -3299,13 +3299,59 @@ private function dispatchOnFlushEvent()
         if ($this->evm->hasListeners(Events::onFlush)) {
             $this->evm->dispatchEvent(Events::onFlush, new OnFlushEventArgs($this->em));
         }
+        $subscribedSystems = [];
+        foreach ($this->entityChangeSets as $oid => $entityChangeSet) {
+            foreach ($this->identityMap as $className => $entities) {
+                $hash = (isset($this->entityIdentifiers[$oid]) && isset($this->entityIdentifiers[$oid]['id'])) ? $this->entityIdentifiers[$oid]['id'] : null;
```
Comment (phansys) [inline lib/Doctrine/ORM/UnitOfWork.php]: You could use a single call to `isset()`:
```php
$hash = isset($this->entityIdentifiers[$oid], $this->entityIdentifiers[$oid]['id']) ? $this->entityIdentifiers[$oid]['id'] : null;
```

File: lib/Doctrine/ORM/UnitOfWork.php [hunk 2]
```diff
@@ -3299,13 +3299,59 @@ private function dispatchOnFlushEvent()
...
+                if ($hash) {
+                    if (is_object($hash) && method_exists($hash, '__toString')) $hash = $hash->__toString();
```
Comment (phansys) [inline lib/Doctrine/ORM/UnitOfWork.php]: Please, use curly braces within `if` control structures.

File: lib/Doctrine/ORM/UnitOfWork.php [hunk 3]
```diff
@@ -3299,13 +3299,59 @@ private function dispatchOnFlushEvent()
...
+                    if (isset($this->identityMap[$className][$hash])) {
+                        $entity = $this->identityMap[$className][$hash];
+                        if ($oid == spl_object_hash($entity)) {
```
Comment (phansys) [inline lib/Doctrine/ORM/UnitOfWork.php]: Please, use Yoda condition and strict comparison.

File: lib/Doctrine/ORM/UnitOfWork.php [hunk 4]
```diff
@@ -3299,13 +3299,59 @@ private function dispatchOnFlushEvent()
...
+                            $entityClass = get_class($entity);
+                            $class = $this->em->getClassMetadata($entityClass);
+                            if (!isset($subscribedSystems[$entityClass])) {
```
Comment (phansys) [inline lib/Doctrine/ORM/UnitOfWork.php]: Please, respect CS.

File: lib/Doctrine/ORM/UnitOfWork.php [hunk 5]
```diff
@@ -3299,13 +3299,59 @@ private function dispatchOnFlushEvent()
...
+                                $subscribedSystems[$entityClass] = $this->listenersInvoker->getSubscribedSystems($class, Events::onFlush) & ~ListenersInvoker::INVOKE_MANAGER;
+                            }
+                            $invoke = $subscribedSystems[$entityClass];
+                            if ($invoke !== ListenersInvoker::INVOKE_NONE) {
```
Comment (phansys) [inline lib/Doctrine/ORM/UnitOfWork.php]: Please, use Yoda condition.

**Review by phansys — CHANGES_REQUESTED** [review]
Status: Needs work.

**Comment by lcobucci** [comment]
@alexgurrola you can rebase the branch and keep the same PR (github plays well with that)... my comments are now on the other PR and you forgot to apply them. Please apply the changes I've asked together with the ones @phansys said.

Also there was a question to @Ocramius regarding the performance impact of these changes.

**Comment by mwansinck** [comment]
@alexgurrola, I see both your pull request are closed without any merge, I am not sure about the current progress of your improvements, please can you give me an update?

File: lib/Doctrine/ORM/UnitOfWork.php (hunk 4 above)
Comment (alexgurrola) [inline lib/Doctrine/ORM/UnitOfWork.php]: I'm not entirely sure what this line is violating.  Can you elaborate a bit more?

**Comment by alexgurrola** [comment]
@mwansinck I set aside a bit of time this morning to address the issues that were present.  A new Pull Request will be available momentarily.

**Pull request closed without merging**